services/dnsaas/zone.go

```diff
@@ -0,0 +1,142 @@
+package dnsaas
+
+import (
+	"context"
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	dnsaas "github.com/ionos-cloud/sdk-go-dnsaas"
+	"github.com/ionos-cloud/terraform-provider-ionoscloud/v6/utils"
+	"log"
+	"strings"
+)
+
+var zoneResourceName = "DNS Zone"
+
+func (c *Client) CreateZone(ctx context.Context, d *schema.ResourceData) (id string, responseInfo utils.ApiResponseInfo, err error) {
+	request := setZoneCreateRequest(d)
+	responseData, apiResponse, err := c.sdkClient.ZonesApi.ZonesPost(ctx).ZoneCreateRequest(*request).Execute()
+	apiResponse.LogInfo()
+	if err != nil {
+		return "", apiResponse, err
+	}
+	return *responseData.Id, apiResponse, err
+}
+
+func (c *Client) IsZoneCreated(ctx context.Context, d *schema.ResourceData) (bool, error) {
+	zoneId := d.Id()
+	zone, _, err := c.GetZoneById(ctx, zoneId)
+	if err != nil {
+		return false, err
+	}
+	if zone.Metadata == nil || zone.Metadata.State == nil {
+		return false, fmt.Errorf("expected metadata, got empty for zone with ID: %s", zoneId)
+	}
+	log.Printf("[DEBUG] zone state: %s", *zone.Metadata.State)
+
+	return strings.EqualFold((string)(*zone.Metadata.State), (string)(dnsaas.CREATED)), nil
+}
+
+func (c *Client) GetZoneById(ctx context.Context, id string) (dnsaas.ZoneResponse, *dnsaas.APIResponse, error) {
+	zone, apiResponse, err := c.sdkClient.ZonesApi.ZonesFindById(ctx, id).Execute()
+	apiResponse.LogInfo()
+	return zone, apiResponse, err
+}
+
+func (c *Client) ListZones(ctx context.Context, filterName string) (dnsaas.ZonesResponse, *dnsaas.APIResponse, error) {
+	request := c.sdkClient.ZonesApi.ZonesGet(ctx)
+	if filterName != "" {
+		request = request.FilterZoneName(filterName)
+	}
+	zones, apiResponse, err := c.sdkClient.ZonesApi.ZonesGetExecute(request)
+	apiResponse.LogInfo()
+	return zones, apiResponse, err
+}
+
+func (c *Client) SetZoneData(d *schema.ResourceData, zone dnsaas.ZoneResponse) error {
+	if zone.Id != nil {
+		d.SetId(*zone.Id)
+	}
+
```

Comment:
if zone.properties == nil - return error

Reply:
@cristiGuranIonos solved in 2273efbf55a163a4f7e5e4a2f4f6bbe854780969 for both zones and records. I also added documentation. If everything else is ok, I can start writing tests.